### ISSUE TITLE
[Request] Support for whisper-ja-anime-v0.3 (Whisper V3 Turbo finetune)

### ISSUE THREAD

**Issue by exn251** [CrispStrobe/CrispASR#258]
whisper-ja-anime-v0.3 is a finetune of whisper v3 turbo focused on Japanese. 
https://huggingface.co/efwkjn/whisper-ja-anime-v0.3

From what I understand here are the issues (LLM summary):

The reason whisper-ja-anime-v0.3 didn't work out-of-the-box on crispasr (and by extension, the original whisper.cpp) comes down to how the model's vocabulary (tokenizer) was modified and how the C++ engine relies on "magic numbers."

Normally, OpenAI's Whisper models use a massive vocabulary of 51,865 tokens to support 100 languages. To make the model run faster and focus solely on Japanese, the creator of v0.3 threw away the standard dictionary and trained a brand new, custom Japanese vocabulary containing only 20,480 tokens.

While this makes the model faster and more accurate for anime, it completely broke crispasr's hardcoded logic in three major ways:
1. The "English-Only" Assumption

Inside the C++ code and the Python conversion scripts, there is a hardcoded rule: If a model has fewer than 51,865 tokens, assume it is an English-only model.
Because this model only has 20,480 tokens, crispasr immediately misclassified it. It completely disabled all language detection, hid the Japanese language tag (<|ja|>), and disabled the translation/transcription task tokens.
2. The "Non-Speech" Suppression Bug (The Gibberish Maker)

To stop Whisper from hallucinating, whisper.cpp runs a hardcoded filter that blocks specific token IDs corresponding to music notes (♪), asterisks (*), and blank spaces in the original OpenAI dictionary.
However, because v0.3 uses a completely different dictionary, those exact same token IDs now represent valid Japanese characters!
By leaving the default filters on, the C++ engine was accidentally censoring random Japanese words and syllables, completely destroying the decoder's ability to form coherent sentences.
3. Missing Special Tokens during Conversion

Standard Whisper models have their special tokens (like <|startoftranscript|> or <|1.50|>) baked directly into the main vocab.json file. Hugging Face fine-tunes often place these in a separate file called added_tokens.json. The legacy convert-h5-to-ggml.py script wasn't programmed to look for added_tokens.json, meaning it exported a binary file that was literally missing the structural tokens required to process audio timestamps.

**Comment by CrispStrobe** [comment]
is its quality any good? can you clarify the license situation? the hf repo specifies none.

anyway, you can just do things like this: 
Use a CrispASR commit that includes issue #258 custom Whisper vocabulary support. The commit used for these artifacts was:

```bash
export CRISPASR_REF=6ba791a5352d
export SRC_REPO=efwkjn/whisper-ja-anime-v0.3
export OUT_REPO=cstr/whisper-ja-anime-v0.3-GGML
export NAME=whisper-ja-anime-v0.3
```

Create an isolated workspace:

```bash
mkdir -p whisper-ja-anime-ggml
cd whisper-ja-anime-ggml
python3 -m venv .venv
. .venv/bin/activate
python -m pip install --upgrade pip
python -m pip install transformers safetensors huggingface_hub tiktoken numpy torch cmake ninja
export USE_TF=0
export HF_HOME="$PWD/hf-home"
export TMPDIR="$PWD/tmp"
mkdir -p "$HF_HOME" "$TMPDIR" source out
```

Clone and build CrispASR:

```bash
git clone --recursive https://github.com/CrispStrobe/CrispASR.git
cd CrispASR
git checkout "$CRISPASR_REF"
git submodule update --init --recursive
cmake -G Ninja -B build -DCMAKE_BUILD_TYPE=Release
export BUILD_JOBS="$(python -c 'import os; print(os.cpu_count() or 2)')"
cmake --build build --target crispasr-cli crispasr-legacy-quantize -j"$BUILD_JOBS"
cd ..
```

Download the Hugging Face source files:

```bash
python - <<'PY'
from pathlib import Path
from huggingface_hub import hf_hub_download
import os

repo = os.environ["SRC_REPO"]
out = Path("source")
out.mkdir(parents=True, exist_ok=True)
files = [
    "config.json",
    "generation_config.json",
    "preprocessor_config.json",
    "tokenizer_config.json",
    "special_tokens_map.json",
    "normalizer.json",
    "tokenizer.json",
    "vocab.json",
    "merges.txt",
    "added_tokens.json",
    "model.safetensors",
]
for filename in files:
    print(f"download {filename}", flush=True)
    path = hf_hub_download(repo_id=repo, filename=filename, local_dir=out)
    print(f"ready {Path(path).name} {Path(path).stat().st_size}", flush=True)
PY
```

Download OpenAI Whisper assets used by the legacy converter:

```bash
git clone --depth 1 https://github.com/openai/whisper.git openai-whisper
```

Convert the HF model to legacy Whisper GGML F16:

```bash
python CrispASR/models/convert-h5-to-ggml.py source openai-whisper out
mv out/ggml-model.bin "out/${NAME}-f16.bin"
```

Quantize with the legacy Whisper quantizer:

```bash
CrispASR/build/bin/crispasr-legacy-quantize "out/${NAME}-f16.bin" "out/${NAME}-q8_0.bin" q8_0
CrispASR/build/bin/crispasr-legacy-quantize "out/${NAME}-f16.bin" "out/${NAME}-q4_k.bin" q4_k
ls -lh out
```

Do not use `crispasr-quantize` for these files. `models/convert-h5-to-ggml.py` produces legacy Whisper GGML files with `lmgg` magic, while `crispasr-quantize` is for GGUF inputs and will reject them with `expected 'GGUF'`.

Smoke-test the Q4_K artifact:

```bash
CrispASR/build/bin/crispasr -m "out/${NAME}-q4_k.bin" -f CrispASR/samples/jfk.wav -l ja --no-timestamps
```

A successful load should report the custom model shape, including:

```text
whisper_model_load: n_vocab       = 20480
whisper_model_load: n_mels        = 128
whisper_model_load: ftype         = 12
```

**Comment by mr3chitk** [comment]
since this was staled, i decided to bump it with his new smaller model: [whisper-ja-760M](https://huggingface.co/efwkjn/whisper-ja-760M)
quality is good. 
- benchmarks are better than final version of cohere ja https://huggingface.co/efwkjn/cohere-asr-ja (which is better than v0.1)
- i tested both cohere ja and whisper-ja-760M. both worked with current scripts, whisper-ja was slower but more stable for sure.
i don't know about licensing thingy but you already had cohere ja v0.1 + whispers here.

also i think --stream-partial-decode-ms doesn't work in v0.8.23 or maybe i read the doc wrongly

LOGS:
Starting translation server in a new window...
Waiting for translation server health endpoint...
Starting crisp-caption bridge...
INFO:__main__:Translation prompt file: system_prompt.txt
INFO:__main__:Glossary file: glossary.json
INFO:__main__:Bridge config: profiles\profile.ja.json
INFO:bridge_runtime:Spawning profile=profile.ja.json: tools/crispasr/crispasr.exe --stream --monitor --no-prints --backend cohere -m C:\Users\PC\crisp-caption\models\asr\cohere-asr-ja-q4_k.gguf -l ja --stream-json --stream-final-mode redecode --stream-utterance-max-sec 60 --stream-final-on-silence-ms 400 --stream-vad-merge-gap-ms 0 --stream-length 10000 --stream-step 1500 --stream-partial-decode-ms 8000 --vad -vm C:\Users\PC\crisp-caption\models\vad\firered-vad.gguf -vt 0.50 -vspd 250 -vsd 250 -vmsd 60 -vp 30 --seed 1 --frequency-penalty 0.2 --beam-size 1 --hotwords-file hotwords.txt --hotwords-boost 2.0 --chunk-seconds 60 --chunk-overlap 0
INFO:bridge_runtime:Queueing initial 48000-byte silence (one Crisp stream step) onto stdin.
INFO:bridge_runtime:Serving http://127.0.0.1:8765/ - select a profile, allow capture, then wait for SDP.
INFO:web_app:WebSocket /ws connected (1 clients)
INFO:web_app:track received kind=audio
INFO:web_app:pc connectionState=connecting
INFO:web_app:pc connectionState=connected
{"type": "transcript", "seq": 1, "kind": "partial", "final": false, "text": "ここの筋替えたけど、これ13", "utterance_id": 1, "t0": 10.39, "t1": 12.0}
{"type": "transcript", "seq": 2, "kind": "partial", "final": false, "text": "ここの筋変えたっけ、べ?これ130でしたっけ?", "utterance_id": 1, "t0": 10.39, "t1": 13.5}
{"type": "transcript", "seq": 3, "kind": "partial", "final": false, "text": "ここの筋変えたっけ?これ130でしたっけ?えー!おろどりって!", "utterance_id": 1, "t0": 10.39, "t1": 15.0}
{"type": "transcript", "seq": 4, "kind": "partial", "final": false, "text": "ここの筋変えたっけ、べこら130でしたっけ?えーー!おろどりって!!1回で4枚も!", "utterance_id": 1, "t0": 10.39, "t1": 16.5}
{"type": "transcript", "seq": 5, "kind": "partial", "final": false, "text": "ここの筋変えたっけべ?これ130でしたっけ?えーー!お呪いって!?1回で4枚も出るんですかー!?", "utterance_id": 1, "t0": 10.39, "t1": 18.0}
{"type": "transcript", "seq": 6, "kind": "partial", "final": false, "text": "ここの筋変えたっけ?これ130でしたっけ?えー!お呪いって!?1回で4枚も出るんですかー!?またパチコキを...", "utterance_id": 1, "t0": 10.39, "t1": 19.5}
{"type": "transcript", "seq": 7, "kind": "partial", "final": false, "text": "これ130でしたっけ?えええっ!お奢りって1回で4枚も出るんですか!?またパチコキをパチコキしかしない!", "utterance_id": 1, "t0": 10.39, "t1": 21.0}
{"type": "transcript", "seq": 8, "kind": "partial", "final": false, "text": "えーっ!?お呪いって1回で4枚も出るんですか!?またパチコキをパチコキしかしないじゃないのこいつマジで。", "utterance_id": 1, "t0": 10.39, "t1": 22.5}
{"type": "transcript", "seq": 9, "kind": "partial", "final": false, "text": "一回で4枚も出るんですかー!?またパチコキを。パチコキしかしねえじゃねえかこいつマジで。パチコキばば。", "utterance_id": 1, "t0": 10.39, "t1": 24.0}
{"type": "transcript", "seq": 10, "kind": "partial", "final": false, "text": "4枚も出るんですかー!またパチコキよ。パチコキしかしねえじゃねえかこいつマジで。パチコキばば。", "utterance_id": 1, "t0": 10.39, "t1": 25.5}
{"type": "transcript", "seq": 11, "kind": "final", "final": true, "text": "ここの筋変えたっけ?これ130でしたっけ?えーー!お呪いって!?1回で4枚も出るんですかー!?またパチコキよ。パチコキしかしねーじゃない のこいつマジで。パチコキばばー", "utterance_id": 1, "t0": 10.39, "t1": 24.22}

SUMMARY:
Spawning profile=profile.ja.json: is a cmd command that is pasted directly to asyncio.create_subprocess_exec with --stream-partial-decode-ms 8000
partial decode still happens every --stream-step 1500

**Comment by CrispStrobe** [comment]
it also has no license?

**Comment by exn251** [comment]
No license stated but looking closer, a lot of the datasets use content ripped from videogames so that is probably why.

**Comment by CrispStrobe** [comment]
Here's how to convert, quantize, and run **whisper-ja-760M** yourself — it's the #258 custom-vocab Whisper path, already in `main`. The source model has **no license**, so build your own GGML locally and don't redistribute it.

**1. Build the converter + legacy quantizer:**
```bash
git clone --recursive https://github.com/CrispStrobe/CrispASR.git && cd CrispASR
cmake -G Ninja -B build -DCMAKE_BUILD_TYPE=Release
cmake --build build -j"$(nproc)" --target crispasr-cli crispasr-legacy-quantize
```

**2. Get the model + the OpenAI mel filters** (the converter reads `whisper/assets/mel_filters.npz`):
```bash
pip install -U "huggingface_hub[cli]" transformers safetensors torch numpy
export USE_TF=0
hf download efwkjn/whisper-ja-760M --local-dir src
mkdir -p wh/whisper/assets
curl -sL -o wh/whisper/assets/mel_filters.npz \
  https://raw.githubusercontent.com/openai/whisper/main/whisper/assets/mel_filters.npz
```

**3. Convert → f16 GGML, then quantize → q4_k:**
```bash
python models/convert-h5-to-ggml.py src wh out            # → out/ggml-model.bin (f16)
build/bin/crispasr-legacy-quantize out/ggml-model.bin out/whisper-ja-760m-q4_k.bin q4_k
```

**4. Run it (Japanese):**
```bash
build/bin/crispasr --backend whisper -m out/whisper-ja-760m-q4_k.bin -l ja -f audio.wav
# (add your --stream/--vad/etc. flags as usual)
```

I verified this end-to-end: the custom 10,240-token vocab converts cleanly (not misclassified as English-only) and the f16 and q4_k transcripts are identical on a JA clip — no visible quant loss — e.g. `って思うかもしれないんですけど 年末年始というと…`. The same recipe works for whisper-ja-anime-v0.3 — just swap the repo id.
